File: benchmark/languages/go/track_a/binary_trees.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

type Node struct {
	item       int
	left,right *Node
}

func makeTree(item int, depth int) *Node {
	if depth == 0 {
		return &Node{item: item}
	}
	childDepth := depth - 1
	return &Node{
		item:  item,
		left:  makeTree(item*2-1, childDepth),
		right: makeTree(item*2, childDepth),
	}
}

func checkTree(node *Node) int {
	if node == nil {
		return 0
	}
	return node.item + checkTree(node.left) - checkTree(node.right)
}

func run(maxDepth int) int {
	minDepth := 4
	stretchDepth := maxDepth + 1

	checksum := checkTree(makeTree(0, stretchDepth))
	longLivedTree := makeTree(0, maxDepth)

	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << (maxDepth - depth + minDepth)
		for i := 1; i <= iterations; i++ {
			checksum += checkTree(makeTree(i, depth))
			checksum += checkTree(makeTree(-i, depth))
		}
	}

	checksum += checkTree(longLivedTree)
	return checksum
}

func main() {
	maxDepth := 12
	if len(os.Args) > 1 {
		if parsed, err := strconv.Atoi(os.Args[1]); err == nil {
			maxDepth = parsed
		}
	}
	fmt.Println(run(maxDepth))
}

File: benchmark/languages/rust/track_a/binary_trees.rs
use std::env;

struct Node {
    item: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn make_tree(item: i32, depth: i32) -> Node {
    if depth == 0 {
        return Node {
            item,
            left: None,
            right: None,
        };
    }
    let child_depth = depth - 1;
    Node {
        item,
        left: Some(Box::new(make_tree(item * 2 - 1, child_depth))),
        right: Some(Box::new(make_tree(item * 2, child_depth))),
    }
}

fn check_tree(node: &Node) -> i32 {
    let left = node.left.as_ref().map_or(0, |n| check_tree(n));
    let right = node.right.as_ref().map_or(0, |n| check_tree(n));
    node.item + left - right
}

fn run(max_depth: i32) -> i32 {
    let min_depth = 4;
    let stretch_depth = max_depth + 1;

    let mut checksum = check_tree(&make_tree(0, stretch_depth));
    let long_lived_tree = make_tree(0, max_depth);

    let mut depth = min_depth;
    while depth <= max_depth {
        let iterations = 1 << (max_depth - depth + min_depth);
        let mut i = 1;
        while i <= iterations {
            checksum += check_tree(&make_tree(i, depth));
            checksum += check_tree(&make_tree(-i, depth));
            i += 1;
        }
        depth += 2;
    }

    checksum + check_tree(&long_lived_tree)
}

fn main() {
    let max_depth = env::args()
        .nth(1)
        .and_then(|s| s.parse::<i32>().ok())
        .unwrap_or(12);
    println!("{}", run(max_depth));
}
